### PR TITLE
Add external_addr to config (network section)

### PR DESCRIPTION
**Pull request by fanatid**
## Motivation

Allow `zebrad` advertise address specified by user.

## Solution

Add `external_addr` to config (network section).

- In the diagram, I see that listen port (address) should be in AddressBook but I can not find code where `listen_addr` added to AddressBook. Can you help with this? https://github.com/ZcashFoundation/zebra/blob/2cecd52a106a8747f2510147481ceb3e372929a7/zebra-network/src/peer_set/candidate_set.rs#L46

The code in this pull request has:
  - [ ] Documentation Comments
  - [ ] Unit Tests and Property Tests

## Related Issues

#1890 

## Follow Up Work

Allow set only IP address in `listen_addr` (default port from `Network` enum).
https://github.com/ZcashFoundation/zebra/blob/2cecd52a106a8747f2510147481ceb3e372929a7/zebra-network/src/peer_set/initialize.rs#L127-L128

### CODE REVIEW

**Comment by teor2345** [comment]
> * In the diagram, I see that listen port (address) should be in AddressBook but I can not find code where `listen_addr` added to AddressBook. Can you help with this?

I think that diagram is a bit confusing, maybe we should say "Inbound Peer Connections to the zebrad listener port".

**Comment by teor2345** [comment]
> > * In the diagram, I see that listen port (address) should be in AddressBook but I can not find code where `listen_addr` added to AddressBook. Can you help with this?
> 
> I think that diagram is a bit confusing, maybe we should say "Inbound Peer Connections to the zebrad listener port".

#2036 might help here, I updated the docs.

**Comment by dconnolly** [comment]
Looks like rustfmt doesn't like the space on one line, causing rustfmt job to fail

File: zebra-network/src/peer/handshake.rs [hunk 1]
```diff
@@ -227,7 +227,10 @@ pub async fn negotiate_version(
         timestamp,
         address_recv: (PeerServices::NODE_NETWORK, *addr),
         // TODO: detect external address (#1893)
-        address_from: (our_services, config.listen_addr),
+        address_from: (
+            our_services,
+            config.external_addr.unwrap_or(config.listen_addr),
```
Comment (teor2345) [inline zebra-network/src/peer/handshake.rs:232]: Can you turn this into an `external_addr()` method on `Config`?
(And make `Config.external_addr` private or `pub(crate)`.)

That way, we'll always get the external address the right way.

(We'll also want to use the external address in #1892 and similar tickets.)

**Review by teor2345 — CHANGES_REQUESTED** [review]
I think we'll want to have separate IPv4 and IPv6 external addresses, and choose the external address based on the connection's IP version. (See #1890.)

We'll also want to:
- add the external address for the connection's IP version to Zebra's `Version` messages
- add both external addresses to Zebra's `Peers` responses
- fallback to the internal address if there is no external address

~~but that depends on #2033 merging:
https://github.com/ZcashFoundation/zebra/pull/2033/files#diff-3f14971a74a109913a4809bd1a77f4e0145e1edfa72aaded8f1fdcb2cc45f4f1R243-R244~~

~~As part of that change, we'll want to make the `AddressBook` ignore any null addresses  (`0.0.0.0` or `[::]`).~~

**Comment by teor2345** [comment]
> I think we'll want to have separate IPv4 and IPv6 external addresses, and choose the external address based on the connection's IP version. (See #1890.)

It might be simpler to allow a list of external addresses, and:
- choose a random one of the correct type for the `Version` response
- put them all in the `Peers` response

That way, we support multi-homed nodes.

I'm not sure if serde will use `to_socket_addrs` to get multiple IP addresses from a single DNS name. We should check what it does and document it.

Edit: supporting multi-homed nodes is optional for now

**Comment by teor2345** [comment]
I'm going to mark this PR as draft until the Zebra team decides which changes we actually need here.

**Comment by teor2345** [comment]
We might merge this as part of ticket #3117 when we get to it.

**Pull request closed without merging**